File: internal/query/cache.go
package query

import (
	"container/list"
	"sync"
	"time"
)

// Cache provides a TTL + LRU-capped in-memory cache for query metadata.
type Cache struct {
	mu      sync.Mutex
	entries map[string]*list.Element
	order   *list.List // front = most recently used
	ttl     time.Duration
	maxSize int
}

type cacheEntry struct {
	key       string
	value     any
	expiresAt time.Time
}

// NewCache creates a Cache with the given TTL and a maximum of 10000 entries.
func NewCache(ttl time.Duration) *Cache {
	return NewCacheWithSize(ttl, 10000)
}

// NewCacheWithSize creates a Cache with the given TTL and max entry count.
func NewCacheWithSize(ttl time.Duration, maxSize int) *Cache {
	return &Cache{
		entries: make(map[string]*list.Element),
		order:   list.New(),
		ttl:     ttl,
		maxSize: maxSize,
	}
}

// Get retrieves a cached value. Returns the value and true if found and not expired.
func (c *Cache) Get(key string) (any, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	el, ok := c.entries[key]
	if !ok {
		return nil, false
	}
	e := el.Value.(*cacheEntry)
	if time.Now().After(e.expiresAt) {
		c.removeLocked(el)
		return nil, false
	}
	c.order.MoveToFront(el)
	return e.value, true
}

// Set stores a value in the cache with the default TTL.
func (c *Cache) Set(key string, value any) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if el, ok := c.entries[key]; ok {
		e := el.Value.(*cacheEntry)
		e.value = value
		e.expiresAt = time.Now().Add(c.ttl)
		c.order.MoveToFront(el)
		return
	}

	// Evict LRU entries if at capacity.
	for c.order.Len() >= c.maxSize {
		c.removeLocked(c.order.Back())
	}

	e := &cacheEntry{key: key, value: value, expiresAt: time.Now().Add(c.ttl)}
	el := c.order.PushFront(e)
	c.entries[key] = el
}

// Delete removes a key from the cache.
func (c *Cache) Delete(key string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if el, ok := c.entries[key]; ok {
		c.removeLocked(el)
	}
}

// Clear removes all entries.
func (c *Cache) Clear() {
	c.mu.Lock()
	c.entries = make(map[string]*list.Element)
	c.order.Init()
	c.mu.Unlock()
}

// GetOrCompute retrieves a cached value, or calls supplier to compute and cache it.
func (c *Cache) GetOrCompute(key string, supplier func() (any, error)) (any, error) {
	if val, ok := c.Get(key); ok {
		return val, nil
	}
	val, err := supplier()
	if err != nil {
		return nil, err
	}
	c.Set(key, val)
	return val, nil
}

// Len returns the number of entries (including expired ones not yet evicted).
func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.order.Len()
}

// EvictExpired removes expired entries.
func (c *Cache) EvictExpired() {
	now := time.Now()
	c.mu.Lock()
	defer c.mu.Unlock()
	for el := c.order.Back(); el != nil; {
		prev := el.Prev()
		e := el.Value.(*cacheEntry)
		if now.After(e.expiresAt) {
			c.removeLocked(el)
		}
		el = prev
	}
}

func (c *Cache) removeLocked(el *list.Element) {
	e := el.Value.(*cacheEntry)
	delete(c.entries, e.key)
	c.order.Remove(el)
}

File: crates/metalog-query/src/cache.rs
use std::{
    collections::HashMap,
    time::{Duration, Instant},
};

/// TTL + LRU cache for query filter rewriting results.
///
/// Entries expire after `ttl` and are evicted when `max_size` is exceeded.
/// Thread-safe via external locking (the caller wraps in `RwLock`).
pub struct Cache {
    entries: HashMap<String, CacheEntry>,
    ttl: Duration,
    max_size: usize,
}

struct CacheEntry {
    value: String,
    inserted_at: Instant,
}

/// Default TTL for cache entries.
const DEFAULT_TTL: Duration = Duration::from_secs(300); // 5 minutes

/// Default maximum entries.
const DEFAULT_MAX_SIZE: usize = 10_000;

impl Cache {
    /// Creates a new cache with default TTL (5 min) and max size (10k).
    pub fn new() -> Self {
        Self {
            entries: HashMap::new(),
            ttl: DEFAULT_TTL,
            max_size: DEFAULT_MAX_SIZE,
        }
    }

    /// Creates a cache with custom TTL and max size.
    pub fn with_config(ttl: Duration, max_size: usize) -> Self {
        Self {
            entries: HashMap::new(),
            ttl,
            max_size,
        }
    }

    /// Gets a cached value if it exists and hasn't expired.
    pub fn get(&self, key: &str) -> Option<&str> {
        let entry = self.entries.get(key)?;
        if entry.inserted_at.elapsed() > self.ttl {
            return None;
        }
        Some(&entry.value)
    }

    /// Inserts or updates a cache entry.
    pub fn set(&mut self, key: String, value: String) {
        if self.entries.len() >= self.max_size && !self.entries.contains_key(&key) {
            self.evict_oldest();
        }
        self.entries.insert(
            key,
            CacheEntry {
                value,
                inserted_at: Instant::now(),
            },
        );
    }

    /// Gets a cached value, or computes and caches it.
    pub fn get_or_compute<F, E>(&mut self, key: &str, compute: F) -> Result<String, E>
    where
        F: FnOnce() -> Result<String, E>, {
        if let Some(val) = self.get(key) {
            return Ok(val.to_string());
        }
        let val = compute()?;
        self.set(key.to_string(), val.clone());
        Ok(val)
    }

    /// Removes expired entries.
    pub fn evict_expired(&mut self) {
        self.entries
            .retain(|_, entry| entry.inserted_at.elapsed() <= self.ttl);
    }

    /// Removes the oldest entry.
    fn evict_oldest(&mut self) {
        if let Some(oldest_key) = self
            .entries
            .iter()
            .min_by_key(|(_, entry)| entry.inserted_at)
            .map(|(key, _)| key.clone())
        {
            self.entries.remove(&oldest_key);
        }
    }

    /// Returns the number of entries.
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    /// Returns true if the cache is empty.
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Clears all entries.
    pub fn clear(&mut self) {
        self.entries.clear();
    }
}

impl Default for Cache {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn set_and_get() {
        let mut cache = Cache::new();
        cache.set("key1".into(), "value1".into());
        assert_eq!(cache.get("key1"), Some("value1"));
    }

    #[test]
    fn missing_key() {
        let cache = Cache::new();
        assert_eq!(cache.get("missing"), None);
    }

    #[test]
    fn expired_entry() {
        let mut cache = Cache::with_config(Duration::from_nanos(1), 100);
        cache.set("key".into(), "val".into());
        std::thread::sleep(Duration::from_millis(1));
        assert_eq!(cache.get("key"), None);
    }

    #[test]
    fn evict_oldest_on_overflow() {
        let mut cache = Cache::with_config(Duration::from_secs(60), 2);
        cache.set("a".into(), "1".into());
        cache.set("b".into(), "2".into());
        cache.set("c".into(), "3".into()); // Should evict "a".
        assert_eq!(cache.len(), 2);
        assert_eq!(cache.get("a"), None);
    }

    #[test]
    fn get_or_compute() {
        let mut cache = Cache::new();
        let val = cache
            .get_or_compute("key", || Ok::<_, String>("computed".into()))
            .unwrap();
        assert_eq!(val, "computed");
        // Second call should return cached value.
        let val2: Result<String, String> =
            cache.get_or_compute("key", || panic!("should not be called"));
        assert_eq!(val2.unwrap(), "computed");
    }

    #[test]
    fn evict_expired() {
        let mut cache = Cache::with_config(Duration::from_nanos(1), 100);
        cache.set("key".into(), "val".into());
        std::thread::sleep(Duration::from_millis(1));
        cache.evict_expired();
        assert!(cache.is_empty());
    }

    #[test]
    fn clear() {
        let mut cache = Cache::new();
        cache.set("a".into(), "1".into());
        cache.set("b".into(), "2".into());
        cache.clear();
        assert!(cache.is_empty());
    }
}
